packages/rgbpp/src/rgbpp/xudt/btc-transfer-all.ts

```diff
@@ -0,0 +1,234 @@
+ */
+export async function buildRgbppTransferAllTxs(params: RgbppTransferAllTxsParams): Promise<RgbppTransferAllTxsResult> {
+  // Prepare base props
+  const maxRgbppCellsPerCkbTx = RGBPP_TX_INPUTS_MAX_LENGTH;
+  const isMainnet = params.isMainnet;
+  const btcSource = params.btc.dataSource;
+  const btcService = btcSource.service;
+  const ckbCollector = params.ckb.collector;
+  const xudtTypeHex = bytes.hexify(
+    blockchain.Script.pack({
+      ...getXudtTypeScript(isMainnet),
+      args: params.ckb.xudtTypeArgs,
+    }),
+  );
+
+  // Get L2 Cells own by the assetAccounts,
+  // and build L1 UTXO IDs (`${txid}:${vout}`) from each cell.cellOutput.lock.args
+  const accountCells = await Promise.all(
+    params.btc.assetAddresses.map((address) => {
+      return limitPromiseBatchSize(() =>
+        btcService.getRgbppAssetsByBtcAddress(address, {
+          type_script: xudtTypeHex,
+        }),
+      );
+    }),
+  );
+  const utxoIds = new Set<string>(
+    accountCells.flat().map((rgbppCell) => {
+      const lockArgs = unpackRgbppLockArgs(rgbppCell.cellOutput.lock.args);
+      return encodeUtxoId(lockArgs.btcTxId, lockArgs.outIndex);
+    }),
+  );
+
+  // Get all L1 UTXOs from the asset accounts
+  const accountUtxos = await Promise.all(
+    params.btc.assetAddresses.map((address) => {
+      return limitPromiseBatchSize(() => btcSource.getUtxos(address));
+    }),
+  );
+
+  // Storage variables for cache querying
+  const invalidUtxoIds = new Set<string>(); // Set<UtxoId>
+  const cellsMap = new Map<string, Cell[]>(); // Map<UtxoId, Cell[]>
+  const utxosMap = new Map<string, Utxo>(); // Map<CellId, Utxo>
+  const cellMap = new Map<string, Cell>(); // Map<CellId, Cell>
+  const utxoMap = accountUtxos.flat().reduce(
+    // Map<UtxoId, Utxo>
+    (map, utxo) => {
+      const utxoId = encodeUtxoId(utxo.txid, utxo.vout);
+      if (utxoIds.has(utxoId)) {
+        map.set(utxoId, utxo);
+      }
+      return map;
+    },
+    new Map<string, Utxo>(),
+  );
+
+  // Get each L1 UTXO's corresponding L2 Cells
+  // If a UTXO has more L2 Cells than maxRgbppCellsPerCkbTx, the UTXO/Cells are invalid.
+  // Invalid L1 UTXO and its corresponding L2 Cells are excluded from the transfer process.
+  const rgbppGroups: { id: string; cells: number }[] = [];
+  await Promise.all(
+    [...utxoIds].map(async (utxoId) => {
+      return limitPromiseBatchSize(async () => {
+        const { txid, vout } = decodeUtxoId(utxoId);
+        const cells = await btcService.getRgbppAssetsByBtcUtxo(txid, vout);
+        if (cells) {
+          cellsMap.set(utxoId, cells);
+        }
+        const utxo = utxoMap.get(utxoId);
+        if (!utxo || !cells || cells.length > maxRgbppCellsPerCkbTx) {
+          invalidUtxoIds.add(utxoId);
+          return;
+        }
+        cells.forEach((cell) => {
+          const cellId = encodeCellId(cell.outPoint!.txHash, cell.outPoint!.index);
```

Comment:
### Discussion
If a BTC UTXO is bound to a DOB cell, maybe it should be added to invalidUtxoIds too?

## Related Test Case
- a BTC UTXO is bound to multi type_scripts (xUDT + DOB)

Reply:
> If a BTC UTXO is bound to a DOB cell, maybe it should be added to invalidUtxoIds too?

I think there is no need to mark as invalid (add to invalidUtxoIds list;

- rgb-sdk will not handle this kind of asset and throw an exception;

- excluded.assets displays > 40(RGBPP_TX_INPUTS_MAX_LENGTH) assets

Reply:
I tested this part with @Dawn-github. If a UTXO is bound to any non-supported type cells, an error will be thrown from the ckb lib. Essentially, if the `codeHash` and `hashType` of the cell's type script do not match the expected values, the type is not supported.
```
Error: The BTC UTXO must not be bound to xUDT and other type cells at the same time
```

I think I need to add a check here.

Reply:
Fixed at cd644176